Memory management cleanup

roystgnr opened the pull request:
Refs #7714

## Reason
There's still quite a lot of manual memory allocation/deallocation in MOOSE, a lot of it unnecessary.  This PR doesn't actually cover the allocation issues that sent me down this rabbit hole, but it should fix up most of the raw pointers and raw allocation that are easy to change.

## Design
* Replace raw (manually allocated/deallocated) pointers with `unique_ptr` where appropriate
* Replace raw `new` with `make_unique` or `make_shared` as appropriate
* Use the "cast reference via a pointer" trick that we use for libMesh mesh iterators, to get rid of the need for (and then get rid of) identical-but-with-more-const copies of pointers in `Assembly`

## Impact
I've tried to limit changes to places where encapsulation allowed me to avoid any public API changes.  There are changes to *protected* member types, though, so it's not impossible that subclasses using those might fail to compile.

roystgnr committed:
std::make_unique in MooseArray

The original code here was fine, but raw new easily becomes a false
positive when hunting for places where it's not fine.

roystgnr committed:
make_unique in PerfNode

roystgnr committed:
_input_parameter_warehouse -> unique_ptr

This doesn't take away our control of when to destruct it, just adds a
layer of safety.

roystgnr committed:
NNL::_secondary_node_range  unique_ptr

roystgnr committed:
Use libMesh add_elem(unique_ptr) API

This gets rid of another raw pointer allocation

roystgnr committed:
SyntaxTree::_root unique_ptr

roystgnr committed:
new -> make_unique

roystgnr committed:
new -> make_unique

roystgnr committed:
new -> make_unique for subpartitioners

roystgnr committed:
new -> make_unique in DataIO

roystgnr committed:
unique_ptr for Abstract Base Class

Saves us a delete, avoids leaking memory if an exception is thrown, etc
etc.

roystgnr committed:
new -> make_shared for MeshFunctions

roystgnr committed:
unique_ptr for serialized solution, residual copy

This gets us to a default destructor, removes redundant bools, avoids
raw allocation, etc.

Refs #7714

roystgnr committed:
One less indirection layer for local_meshfuns

And this gets rid of some manual memory management too.

roystgnr committed:
unique_ptr in SyntaxTree containers

This makes the code a lot cleaner

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1346648/) on 3a68d58 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23409/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23409/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d6f8146eb254e05879795a1f1bf0ab004e1b72f9`

GiudGiud commented on framework/src/transfers/MultiAppGeneralFieldShapeEvaluationTransfer.C:
```diff
@@ -72,11 +72,11 @@ MultiAppGeneralFieldShapeEvaluationTransfer::buildMeshFunctions(
-                                     from_sys.get_dof_map(),
-                                     from_var_num));
+    std::shared_ptr<MeshFunction> from_func =
+        std::make_shared<MeshFunction>(getEquationSystem(from_problem, _displaced_source_mesh),
```
same

roystgnr commented on framework/src/transfers/MultiAppShapeEvaluationTransfer.C:
```diff
@@ -245,11 +245,11 @@ MultiAppShapeEvaluationTransfer::transferVariable(unsigned int i)
-                                     from_sys.get_dof_map(),
-                                     from_var_num));
+    std::shared_ptr<MeshFunction> from_func =
+        std::make_shared<MeshFunction>(getEquationSystem(from_problem, _displaced_source_mesh),
```
Trying to make minimal changes, I guess, but you're right; there's no need for a smart pointer here at all.

roystgnr committed:
Get rid of a layer of indirection

Thanks to Guillaume for pointing these out

roystgnr commented:
Bah.  Is 14.0.0 not a safe version of clang-format for me to use?  I thought I was being clever by finally just auto-running that when saving files.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1347234/) on 395254d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23409/site)

This comment will be updated on new commits.

roystgnr committed:
unique_ptr, "constify_ref" in Assembly

Using std::unique_ptr here cleans up the destructor and makes the code
clearer ... where we can get away with it.  We can't get a
reference-to-pointer from a unique_ptr, so we have to maintain a ton of
raw pointers to be able to supply those APIs.

The "constify_ref" trick is a much cleaner way of convincing a C++
compiler to give us a T * const & into a const T * const &; it really is
the same address underneath, so we don't need to maintain a copy of
everything, we just need to convince the compiler that it doesn't need
to try to make a temporary copy either.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1348238/) on 73b275e wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>afc31d</th>
      <th colspan=4>#23409 73b275</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.72%</td>
      <td>84.71%</td>
      <td>-0.01%</td>
      <td>73.95%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>82942</td>
      <td>82888</td>
      <td>-54</td>
      <td>247</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14960</td>
      <td>14958</td>
      <td>-2</td>
      <td>87</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23409/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23409/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23409/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 73.95% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud reviewed (commented):
added the valgrind recipes

GiudGiud reviewed (approved):
nice cleanup. This will surely help us do better at memory management by showing the better practices
